### DISCUSSION TITLE
Axe update

### DISCUSSION THREAD

**Discussion by jsaintauret** [marcomusy/vedo#586] [Q&A]
Hello everyone!

I am working on a post-processing tool to visualize data from chemical analysis. At the moment I'm using Pyvista to visualize data as 3D mesh surface plot. To make it short and to give you a quick overlook at what the tool does, is that I create with NumPy three X , Y , Z matrices of size nxm with the meshgrid function then feed these to the StructuredGrid(X,Y,Z) function of pyvista to create a mesh for the plotter(StructuredGrid() of pyvista instantiate a vtkStructuredGrid then sets its point a vtkPoints object).

I came across vedo and was surprised to see that a bunch of features that I was looking for was already available there as opposed to Pyvista (this Is not a shot to pyvista at all, it is a great tool but in the long run vedo looks more suitable for my needs, so I'm thinking on migrating to it).

I do have a few questions though regarding a similar function StructuredGrid() of pyvista to create a 3D surface plot mesh with 3 matrices. I did look through a lot of the examples but it looks like the Volume() and Mesh() instances do not allow me to do it. 
Any suggestion on that side ?  (right now what am I doing is to create a vtkStructuredGrid() object with my X,Y,Z matrice)

The other question that I have is that I'd like to know if there is a way to set/update the Axe() object attribute on the fly since I will embedded the vedo plotter in a qt window? The only way that I seed is to create a callback function from the plotter and then overwrite the current Axe() actor on the plot with a new Axe() object with the attribute that I want.

Thank you a lot in advance for your help and sorry for the long message.

**Comment by marcomusy** [comment]
Hi @jsaintauret at present `vedo` is not directly supporting vtkStructuredGrid dataset (maybe in the future it will!), yet if you already have a pyvista.StructuredGrid or vtkStructuredGrid object **and** you only aim at post-processing visualisation you can still convert it to a vedo object simply with `vedo.Mesh(my_struct_grid)`. Also, consider that `vedo` and `pyvista` are largely interoperable! So I see 2 options:
1. _use vedo with pyvista_: you can create/manipulate your dataset in pyvista and then use it in the vedo environment for visualisation
2. _use pyvista with vedo_: use what you already have in pyvista but add features from vedo like the `Axes` object  which I think you can add in a pyvista scene with `pv.plotter.add_actor(my_vedo_axes)`. E.g. as discussed in [here](https://github.com/pyvista/pyvista/issues/796) you can do something like:
![](https://user-images.githubusercontent.com/32848391/88178051-13236180-cc2a-11ea-87e5-f05ab680df00.png)

**Reply by jsaintauret** [reply]
Hey @marcomusy thanks a lot for your answers , I was actually thinking of using pyvista and vedo together but I was not sure of the efficiency.

Using the vedo actor in pyvista could work for the Axe, I think I still need to see which one of vedo or pyvista gives me better option, but yes working with both might work.

And what about the last part of my question regarding changing the axe attribute on the fly or maybe what I exposed is the only available way for now?

**Comment by marcomusy** [comment]
hi - sorry I overlooked your second question..

> The other question that I have is that I'd like to know if there is a way to set/update the Axe() object attribute on the fly since I will embedded the vedo plotter in a qt window? The only way that I seed is to create a callback function from the plotter and then overwrite the current Axe() actor on the plot with a new Axe() object with the attribute that I want.

yes - in a callback function you can access the `Axes` object and modify sub-parts that already exist (you do this with `myaxes.unpack()`).
Most probably you want to modify it completely... in that case just create a brand new one (you do this by passing a pyvista object or the individual ranges e.g. `Axes(xrange=(0,1))` etc).

**Reply by jsaintauret** [reply]
Hey thank for your answer,after digging a bit into vedo source code I'm starting to understand the philosophy.
I think one other way to do it is to get the actors associated to the current renderer and get the Assembly of the Axe and then unpack it or use the GetParts method of the vtkAssembly to finally update the parameter of a specific element that of the Axe. It might be easier to just recreate a new Axe though and overwrite it in the plotter like you mentioned.

Anyway thank you again for all the information!

**Reply by marcomusy** [reply]
yes - but even easier: you can keep track of the created object in a class or as global variable and overwrite it in a loop or callback, or unpack it with `myaxes.unpack()` to access all the subelements.
Ceck out `examples/pyplot/customIndividualAxes.py`
Let me know if need help on that.

**Reply by jsaintauret** [reply]
Ok I'll take a look at the example.

What I'm planning to do is to develop a toolbar or a qdialog window to customize every aspect of my plot like the Axe parameters, the color map, the axis scale size ect , the scalar bar range.

What I have in mind is something like that https://fr.mathworks.com/help/matlab/creating_plots/customize-figure-before-saving.html or https://fr.mathworks.com/help/matlab/ref/figure.html
It could be good to have something like the matlab figure menu and toolbar to customize the look of the renderer.

I'll keep you updated on my work if you are interested.

**Reply by jsaintauret** [reply]
Hello, it looks like the Base3DProp class is using methods from the vtkProp3D class but I don't see it inheriting from this class since it is instantiated as follow Base3DProp(**object**). Could you please explain how it has access to the method as **self.SetUserTransform**,**self.ComputeMatrix** ,**self.RotateWXYZ** ect ? I'm asking that because I want to know all the property I can access from an actor.

Thanks

**Reply by marcomusy** [reply]
Hi, indeed it's a good question because the Base3DProp instance acquires access only later when Actors and Volumes are created, so at least the name is a bit misleading, I need to see if there is a better way to achieve the same..

**Comment by jsaintauret** [comment]
Hello, 

I'm having a hard time figuring how to update the axes range value of my mesh.
![image](https://user-images.githubusercontent.com/52704901/156212410-6f7bfd07-c426-41ec-93d9-d75f5f3324cc.png)

In the picture, you can see that the range of my mesh for the x,y,z axis is [0,1000]. For the z-axis it is perfectly fine because my data values is in that range, but since the x and y value are very small compared to the z value (x range [0,16] | y range [0,44] ) I had to scale all the x and y coordinate point of my mesh as follow, so the render could be readable :
![image](https://user-images.githubusercontent.com/52704901/156213261-165c187a-0cec-49ef-9cea-9c4e83dc8519.png)

I would like to know if there is a way of changing the range value displayed on the x-axis and y-axis in the same way addElevationScalars would do on the Z-axis ?

I've tried to play with the xrange and yrange arguments of the Axes class but from my understanding, it only changes the size of the axes line according to the defined range selected.

On Pyvista you can use the show_bounds method of the plotter to override the mesh bounds in the form [xmin, xmax, ymin, ymax, zmin, zmax]. https://docs.pyvista.org/api/plotting/_autosummary/pyvista.Plotter.show_bounds.html
What it does is that it calls the setBounds() of the vtkCubeAxesActor.

Thanks

**Reply by marcomusy** [reply]
Hi!
You cannot "update" the range of one of the axes, you need to recreate the whole object and discard the old one.
Just pass the object to the `Axes(my_mesh)` or set manually the bounds with `Axes(xrange=(), ...)`.

But you seem to have tried that already?

> I've tried to play with the xrange and yrange arguments of the Axes class but from my understanding, it only changes the size of the axes line according to the defined range selected.

yes.... what else should it be doing ? :)

**Reply by marcomusy** [reply]
...or maybe you mean that you want to show on the x-axis values [0,16] ?

**Reply by jsaintauret** [reply]
Thanks for your answer!

I hope you didn't take it the wrong way when I say it "only changes" it was more just my understanding from that function ;)

Well, I don't think it should be doing something else since it is how you design it, but I was expecting it to only change the range value display on the axes and not the size.

Do you have any recommendations to keep the range value of my raw data if I rescale all my points in a sort that the plot looks readable?

This is how the mesh looks if I remove the scaling part on the mesh points
![image](https://user-images.githubusercontent.com/52704901/156218079-b4f622e3-85f1-4265-9a84-cbac743cdd1f.png)
![image](https://user-images.githubusercontent.com/52704901/156218087-fb8cc5e1-2a81-4244-869d-e896e5713307.png)

**Reply by marcomusy** [reply]
> I hope you didn't take it the wrong way when I say it "only changes" it was more just my understanding from that function ;)

no worries! LOL

I maybe see what you mean.. you would like to see the range [0,16] on the y-axis and [0,1000] on the z-axis, but the physical size should be about the same?

**Reply by jsaintauret** [reply]
Ok cool :)

Yes on the first picture where the mesh is scaled and looks right, I would like to see the range [0,16] on x , [0,44] on y and [0,1000] on z .

**Reply by marcomusy** [reply]
OK. Now I understand.

I dont' think that you can do it in vtk nor in pyvista with SetBounds()...

In any case you need to explicitly construct the labels of your axis to achieve that, this is possible with the `xValuesAndLabels` keyword. 
Keep in mind that in vedo/vtk/pyvista all it's rendered is in 3D world coordinates.
Check out: `examples/pyplot/customAxes1.py`

That said, _I would not do it_ unless it's really necessary... wouldn't it be better to scale down the z axis and declare the scaling factor in the label itself (eg. "z /scale")?

**Reply by jsaintauret** [reply]
> OK. Now I understand.
> 
> I dont' think that you can do it in vtk nor in pyvista with SetBounds()...
> Ok from what I remember it does work on pyvista when I use to change the bound range but I'm doubting now that you are saying that :/
> In any case you need to explicitly construct the labels of your axis to achieve that, this is possible with the `xValuesAndLabels` keyword. Keep in mind that in vedo/vtk/pyvista all it's rendered is in 3D world coordinates. Check out: `examples/pyplot/customAxes1.py`
Ok I'll take a look at it and see if I come out with something and thanks for the hint
> 
> That said, _I would not do it_ unless it's really necessary... wouldn't it be better to scale down the z axis and declare the scaling factor in the label itself (eg. "z /scale")?
I was actually thinking of doing that instead of scaling the x and y but the thing that I like is to have a plot with bounds that have the symmetric shape of a cube boxe ,like having normalized axes.

**Reply by jsaintauret** [reply]
> That said, _I would not do it_ unless it's really necessary... wouldn't it be better to scale down the z axis and declare the scaling factor in the label itself (eg. "z /scale")?

are you talking about this argument in the Axes class? : textScale, [1.0], global scaling factor for all text elements (titles, labels)

**Reply by marcomusy** [reply]
No, that just scale fonts.
Simply divide your z-values by a scale and declare it in the axis label eg:
`ztitle=f"z, measrued in units of \mum /{my_fave_scale}"`.

**Reply by jsaintauret** [reply]
> In any case you need to explicitly construct the labels of your axis to achieve that, this is possible with the `xValuesAndLabels` keyword. Keep in mind that in vedo/vtk/pyvista all it's rendered is in 3D world coordinates. Check out: `examples/pyplot/customAxes1.py`

Hey just to give you some feedback, even though you didn't recommend it to explicitly construct the labels of the axis, I found it to be the best solution that I was looking for. Your proposition to only mention the scaling factor in the ztitle argument on the Axe was not really what I was looking for.

Here is how I did it:
![image](https://user-images.githubusercontent.com/52704901/156556140-0ef2cd1d-5ca6-44df-a3a9-27c0f09c846f.png)
It is like a mapping of the axis 3D coordinate to the real scalar value of the plot, I'll probably make a function of for it

here are the results:
with 5 points
![image](https://user-images.githubusercontent.com/52704901/156556323-3b8801ba-b78c-4baa-a506-324d85531dcf.png)


with 10 points
![image](https://user-images.githubusercontent.com/52704901/156556517-9f1ae7ae-a5bf-4844-b6f6-54f3d0a0af0d.png)


One more question, I see that the axe doesn't display the last value of the range (xmax,ymax,zmax) , I checked the arguments but didn't find anything is there any reason to that?

thanks for your time

**Reply by marcomusy** [reply]
Nice !
the last tick value is always dropped, so you need to put it twice in your list if you want it.
there are also utilities functions which might help: `utils.precision()` and `utils.linInterpolate(x, rangeX, rangeY)` to map ranges.
I would suggest to remove the non significant zero digits and maybe restore perspective view (parallel view always looks a bit fishy to me..), you can also change fonts to make it look better :)
you can also play with lighting , eg `mesh.lighting('glossy')` .. etc
Thanks for your feedback

**Reply by jsaintauret** [reply]
Thanks also for your feedbacks! I'll take a look at it and I'll let you know

**Reply by jsaintauret** [reply]
Hello,

So I've tried to use `utils.linInterpolate(x, rangeX, rangeY)` to map the range of my axes as you told me.
![image](https://user-images.githubusercontent.com/52704901/163425843-1cfb4282-443d-4a00-b645-2d3a72a780a5.png)
![image](https://user-images.githubusercontent.com/52704901/163425903-0118e1f3-b7ef-41c6-9fe8-b9332cee4d39.png)
 
However I used the `utils.roundToDigit()` instead of `utils.precision()` , but I'm not sure which of those two I need to use to remove the non significant zero digits.

I also have another issue when I want to update the property of my axe, and as you told me before the easiest way to update the axe is to create another axe assembly. So I've tried something by adding a button to update my axe with a new one when I click on it the `udpatheMesh()` method is called.
![image](https://user-images.githubusercontent.com/52704901/163426968-e0459c96-1055-4e00-9db4-5dd0a6db6610.png)

But it looks like the previous axe is not overridden by the new one
![image](https://user-images.githubusercontent.com/52704901/163427459-bdbaa330-ec9b-49b4-bd87-9d068e52c925.png)

I'm not sure to understand how the actors are updated in the plotter. Could you please told me the proper way to replace the old axe with a new one ? I was thinking of removing the Axe assembly from the plotter axes list , but after debugging I didn't find the old axe only the new one:

Debug (break point after first axe )
![image](https://user-images.githubusercontent.com/52704901/163429136-958af50d-3b4c-4a2d-ab5e-dbfe850de01e.png)

Debug (break point after new axe)
![image](https://user-images.githubusercontent.com/52704901/163429436-9538ad7c-0e0a-48ad-ba4b-5865125312af.png)

You can see that the second vtkAssembly adress is not the same as the first , so it should be only the new one display on the renderer right?

**Comment by marcomusy** [comment]
In this example you can see how to remove and update the axes: you should see a sphere "walking" in x-y while the axes are updated (ignore the timer mechanism if you don't need it):
```python
import time
from vedo import Plotter, Sphere, Axes


def bfunc():
    global timerId
    plotter.timerCallback("destroy", timerId)
    if "Play" in button.status():
        timerId = plotter.timerCallback("create", dt=10)
    button.switch()

def handle_timer(event):
    t = time.time() - t0 
    sph.pos(t,t)
    axe = Axes(sph)
    # remove the old and add the new one
    plotter.remove("Axes").add(axe).resetCamera()


timerId = None
t0 = time.time()
plotter= Plotter()
button = plotter.addButton(bfunc, states=[" Play ","Pause"], size=40)
evntId = plotter.addCallback("timer", handle_timer)

sph = Sphere()
plotter += sph
plotter.show().close()
```

<img width="458" alt="Screenshot 2022-04-14 at 20 47 51" src="https://user-images.githubusercontent.com/32848391/163457007-88e12272-82c4-4370-8de6-8387041f9baa.png">

**Reply by jsaintauret** [reply]
Perfect! this was exactly what I was looking for thanks
![image](https://user-images.githubusercontent.com/52704901/163539368-f6bf75e5-d719-4f97-b70e-a11115b8de34.png)

**Reply by jsaintauret** [reply]
Hi Marco, 
sorry again to bother you and I realy hope you don't mind me asking a lot of question. Believe me when I say that before asking you, I try my best first to look on other discussion or vedo documentation where I usually find answers!

When I rescale my mesh after I clipped it and add it back to the render scene with the following:
```
mesh.scale(s=(sx,sy,sz))
self.vp.remove(selected_analysis.sheet_name).add(mesh)
```
I'm expecting to have my mesh align to the axes origin, however when I debug and watch the `axes.origin()` the value is à 0,0,0 as for the mesh origin. Also when I check the `axes.pos()` and `mesh.pos()` it looks like the coordinates are also 0,0,0 but the .bound() value return different value. Could you please give me information about the render scene coordinate system? what value the pos, origin and bound return? how actors are positioned relative to each other (I know that I can use the `.alignTo` function to align meshes)
![image](https://user-images.githubusercontent.com/52704901/163730060-0486c26e-201e-4393-8514-833a2166e7a9.png)

**Comment by marcomusy** [comment]
>  I realy hope you don't mind me asking a lot of question

not at all, no worries!

It looks to that your mesh genuinely starts at about x=25 in the world coordinates, while you are creating the axes in a different range.

Forget the `origin()` method! that has a completely different meaning (it's an unfortunate coincidence that `Àxes` class inherits it with that name...)
Indeed `.bounds()` is what you should look at.

So i'm not sure what could be the problem... I cannot make a more informative guess..

**Comment by marcomusy** [comment]
ps
why are you scaling in sx and sy? I understood that you only needed to scale the vertical axis.

**Reply by jsaintauret** [reply]
Ok, let me explain what I am trying to do. 

Let say for instance that I would like to clip my mesh on the three different axes (x,y and z) with a specific range. So let's take a simple case that I'd like to clip mesh on the x axis on the range [8:16]. When the clipping is done what I want to do is to rescale the clipped mesh to the Axes shape so to do that I compute a scaling factor based on the axes bound on the new bound of the clipped mesh and then apply the scaling factor on sx sy and sz.

See the attached video that shows you what I'm trying to do , the first part is my tool and the second part is an other tool that doesn't what I'm trying to mimic ( I will have to delete the video after because i should share online the second tool)

**Reply by jsaintauret** [reply]
The issue I have is when I apply a scaling factor on the clipped mesh , its position shift also so I need to position it to the axe origin.

I could build a new axe to it but I'm not a big fan of that solution I think I should shift is position to the original axes which is fixed (maybe I'm over-design the solution idk...)

**Reply by marcomusy** [reply]
Oh well I'm not sure `vedo` is well designed for this!
While this is **technically possible** (in fact this is how the `pyplot.py` submodule internally works) it might become a bit tricky. The scope of vedo is to analyse and visualise objects in world coordinates.. if you scale objects , the Axes cannot know the difference and will still represent world coordinates.
What you basically want is to map world coords of a Mesh that you already have, to a R3 space with predefined aspect ratio (e.g. I decide that my x axis spans a length of 2, my y-axis spans 3 units and my z-axis spans 4 units and everything must stretch to this sizes).

If you wish to give it a try you will have to build the axes in more "manual" way, as in:
https://github.com/marcomusy/vedo/blob/cdc1d8cad8e56280d019ca62c7555c9bf936302a/vedo/pyplot.py#L167
and when cutting your Mesh you need remapping the vertices onto this R3 space.

this may sound very complicated but maybe is not that complicated :)

**Reply by jsaintauret** [reply]
Ok I think I'm kinda get a it! and yes indeed I get that vedo is not design for this kind of need, actually I should just use maplotlib which everything already built in , but performance wise using vedo/vtk is so much better and the look and feel of 3D mesh worth the try to find a workaround, and to be honest it's kind of a challenge to make it work I enjoy it. 
Well thank for your info, it might or might not be complicated , but since vedo is well documented and has already a lot of features I'm pretty confident I just got to be oriented in the good direction ;)

**Reply by marcomusy** [reply]
I like your optimism! Let me know if you get stuck anywhere. What I would try:
- make a working copy of your mesh
- cut it the way you like
- map its bounding box to a new fixed bounding box at origin with the desired aspect ratio, with `transformWithLandmarks()`as in `examples/basic/align5.py`
- For the axes: use `utils.makeTicks()` to build them and set them with `axesopts["yValuesAndLabels"] = labs` as above.

it should not be difficult after all!

**Reply by jsaintauret** [reply]
Ok thank you I'll try that and let you know.
One more thing that you could clarify again when you mean it's a world coordinate, is there actually a origin position for the world coordinate like 0,0,0 ? Or each mesh has its own relative origin ? And could you specify more the coordinate value return by the bound function ? 
Thanks

**Reply by marcomusy** [reply]
- yes
- no
- method `.bounds()` returns [x0,x1, y0,y1, z0,z1] in the world c.s.

**Comment by jsaintauret** [comment]
Hi,
so I've used `mesh.alignToBoundingBox()` which use `transformWithLandmarks()` and I've got what I was looking for, it is easier that way than using the mesh bounds to get the cordinate and shift the mesh back to the axes.

Thanks again

https://user-images.githubusercontent.com/52704901/164001836-53a3c97f-693b-4dcb-a260-57e77b3a1f2a.mp4

**Reply by marcomusy** [reply]
Looks great, I didn't think of `alignToBoundingBox`! You're now officially a `vedo` expert LOL
![medal](https://user-images.githubusercontent.com/32848391/164003099-1d5dec93-f47d-4033-936c-62e4e3105aef.png)

**Reply by jsaintauret** [reply]
Haha I humbly accept your medal ;) but I got to say that the documentation is the real MVP !
However I was kinda expecting to not have to scale the clipped mesh prior using the `alignToBoundingBox` since the rigid argument is False, it would have save me some line of code with all the scaling and everything but hey no need to be to greedy !

**Comment by marcomusy** [comment]
> I was kinda expecting to not have to scale the clipped mesh

You are right! Infact I will change the default behavior of this method.... for the next release!

**Reply by jsaintauret** [reply]
actually it didn't not scale as I was expecting it to, since the default value is rigid=False

**Reply by marcomusy** [reply]
Yes. because the scaling is the same in all 3 directions.. while this will be fixed in the next release you can add this small function to replace it temporarily:
```python
    import vtk
    def alignToBoundingBox(source, target, rigid=False):
        lmt = vtk.vtkLandmarkTransform()
        ss = vtk.vtkPoints()
        xss0,xss1,yss0,yss1,zss0,zss1 = source.bounds()
        for p in [[xss0, yss0, zss0],
                  [xss1, yss0, zss0],
                  [xss1, yss1, zss0],
                  [xss0, yss1, zss0],
                  [xss0, yss0, zss1],
                  [xss1, yss0, zss1],
                  [xss1, yss1, zss1],
                  [xss0, yss1, zss1],
                 ]:
            ss.InsertNextPoint(p)
        st = vtk.vtkPoints()
        xst0,xst1,yst0,yst1,zst0,zst1 = target.bounds()
        for p in [[xst0, yst0, zst0],
                  [xst1, yst0, zst0],
                  [xst1, yst1, zst0],
                  [xst0, yst1, zst0],
                  [xst0, yst0, zst1],
                  [xst1, yst0, zst1],
                  [xst1, yst1, zst1],
                  [xst0, yst1, zst1],
                 ]:
            st.InsertNextPoint(p)

        lmt.SetSourceLandmarks(ss)
        lmt.SetTargetLandmarks(st)
        lmt.SetModeToAffine()
        if rigid:
            lmt.SetModeToRigidBody()
        lmt.Update()
        source.applyTransform(lmt)
```

**Reply by jsaintauret** [reply]
Just implemented it and it works perfectly ;).  I'll be waiting on the next release then.

**Comment by jsaintauret** [comment]
Hello Marco,

Here is a  quick update on what I was trying to do with the axe update, I manage to do something pretty cool so here a small demo clip of my app (I'm not set yet on the GUI part for the plotter/axe property option)

https://user-images.githubusercontent.com/52704901/168466296-1dd9116e-d5e8-47f7-9a3d-641ee8630b37.mp4

I do have a question though, is there a way to place the y axis line at the opposite side ? like what we can do with the grid plane (yzGrid2).

By the way you told me that the last tick value is always dropped, so I  need to put it twice in my list . I did that but it doesn't seems to change anything.

**Reply by marcomusy** [reply]
Looks very cool!
Have you tried `yShiftAlongX` in
https://vedo.embl.es/autodocs/content/vedo/addons.html#vedo.addons.Axes

check out also:
`examples/volumetric/earth_model.py`

> By the way you told me that the last tick value is always dropped, so I need to put it twice in my list . I did that but it doesn't seems to change anything.

uhm not sure about that, i need to check..

Btw I noticed in you video that some of the z-scale labels are missing at minute 0:17, is that intended?

Also you may play with camera positioning and zooming to avoid parallel projection and create a better intuition of perspective depth.

**Reply by marcomusy** [reply]
another piece of advice i have: if you know your "zero" level (it looks like you have a flat region defined) then you can use a divergent color mapping which would look more informative that a simple z-coloring.

**Reply by jsaintauret** [reply]
> Looks very cool! Have you tried `yShiftAlongX` in https://vedo.embl.es/autodocs/content/vedo/addons.html#vedo.addons.Axes
>
Perfect thanks ! I will also have to put the tick label outside of the plane to

> Btw I noticed in you video that some of the z-scale labels are missing at minute 0:17, is that intended?
> 
Yes good catch, I saw that to, it is not intended at all. I need to dig in to that to understand what it's happening.

> Also you may play with camera positioning and zooming to avoid parallel projection and create a better intuition of perspective depth.
> 
I remember you mentioning that the parallel projection looks sketchy to you. My tool is still at a early stage so I'm still playing around with everything and explore option, but I'll definitely give the user the liberty to activate or no as displayed on my small clip (sometime parallel projection when setting the camera on the top is better)


https://user-images.githubusercontent.com/52704901/168629924-ac5fe523-993a-4fe2-b637-0d1648916f6c.mp4



However, I've encountered issue when I activate the parallel projection, it is like the zoom is changing when I toggle the option. It is like the parallel projection has its own zoom. I checked the camera position and angle and its stays the same , I don't know where to get the zoom value.

As for your last quote  regarding the colormap I'm not set on it. I just threw a random cmap, the user will have to select his cmap with a combobox for instance and maybe I will try to implement something like a cmap editor.

**Reply by marcomusy** [reply]
hi, nice! probably you are looking for SetParallelScale()
https://vtk.org/doc/nightly/html/classvtkCamera.html#ad0054858ddff1c0b142ebe289d7f4c43

you use it like this:
`plt.camera.SetParallelScale(1.0)`

**Reply by jsaintauret** [reply]
So I actually did try to monitor with GetParallelScale and the return value before and after the parallel projection did not change

![image](https://user-images.githubusercontent.com/52704901/168646851-cb3a4a68-e746-4d00-942d-3364625ed0b8.png)

**Reply by marcomusy** [reply]
uhm you might have a look at
https://discourse.vtk.org/t/enable-parallel-projection-without-resetting-the-camera/5289/7
infact i could implement that solution as the default

**Reply by jsaintauret** [reply]
> uhm you might have a look at https://discourse.vtk.org/t/enable-parallel-projection-without-resetting-the-camera/5289/7 infact i could implement that solution as the default
>

Perfect thanks, it works!
![ezgif com-gif-maker](https://user-images.githubusercontent.com/52704901/168823302-3105f108-43aa-4d9a-a5ae-6f7dcaa52c8c.gif)

**Reply by marcomusy** [reply]
very nice!!!

**Comment by jsaintauret** [comment]
Hi

I didn't realize that I was working with vedo 2022.1.0 so I updated to the last revision and something weird happened. 

![image](https://user-images.githubusercontent.com/52704901/170028941-d84552d3-87b8-4ffe-aeb3-71659c97e266.png)

Did you change anything on regarding the Mesh class ?

**Reply by marcomusy** [reply]
that has changed to `Axes(my_mesh)`

**Reply by jsaintauret** [reply]
Ok I get it thanks, I'll make some changes.

Is it possible to attach a signal from a widget as event for the call back function ?

**Reply by marcomusy** [reply]
what widget are you using?

**Reply by jsaintauret** [reply]
It depends, it goes from simple QPushButton to a QDoubleRangeSlider (which is a custom widget from https://github.com/napari
[superqt](https://github.com/napari/superqt)

![image](https://user-images.githubusercontent.com/52704901/170040295-1d51ecc9-c449-4a90-8afa-73a3c9e1d31a.png)

**Reply by marcomusy** [reply]
uhm but that is not related to vedo (?)
in your class from your qt whatever member function you can access the vedo objects you create.

**Reply by jsaintauret** [reply]
Yes it is not related to vedo sorry for the confusion.

My question was more to see If I could use the add_callback framework of the plotter to execute action while the show() is active. 

When I want to update on the fly a property of a mesh following a widget signaI, what I usually do is:

- get the mesh from the plotter, 
- update it 
- and then I use the render() method from the plotter to update the scene or use plotter.add() and plotter.remove()

**Reply by marcomusy** [reply]
yes - its' probably possible - I would try to call `plotter.addCallback()` before the `show()`.
You should have in your class a `self.plotter = vedo.Plotter()` that you can access at will in your class.

**Reply by jsaintauret** [reply]
Sorry I think I am not really clear.

as per your example, to add a callback on your plotter you do the following `plt.addCallback('mouse move', func) ` with func being the function to excute and 'mouse move' being an event from the mouse.

I was wondering if the event could be for instance:
  `checkbox = QCheckBox()`
   `plt.addCallback(checkbox.checked, func) `

**Reply by marcomusy** [reply]
No! You cannot do that.. but you can call anything from inside the `func`.
